### PR TITLE
avoid holding memory between events in Multi5x5ClusterProducer 

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- modernized Multi5x5ClusterProducer
  - added fillDescriptions
- reworked Multi5x5ClusterAlgo
  - removed use of member data just for temporary storage
  - moved code doing related things into new functions/classes
  - use more modern C++ constructs

The holding and then clearing of memory across events was found by a prototype Alloc Monitor service.

NOTE: the only reason I could not change Multi5x5ClusterProducer to be a global module was the use of `PositionCalc` in Multi5x5ClusterAlgo.

#### PR validation:

Code compiles and running a very simple RECO job succeeded.

This is purely a refactoring so should not affect any physics results.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: RecoEcal/EgammaClusterProducers/src/Multi5x5ClusterProducer.cc [hunk 1]
```diff
@@ -34,142 +36,112 @@ class Multi5x5ClusterProducer : public edm::stream::EDProducer<> {
...
+  edm::ParameterSetDescription posCal;
+  posCal.add<bool>("LogWeighted");
+  posCal.add<double>("T0_barl");
+  posCal.add<double>("T0_endc");
+  posCal.add<double>("T0_endcPresh");
+  posCal.add<double>("W0");
+  posCal.add<double>("X0");
+  ps.add<edm::ParameterSetDescription>("posCalcParameters", posCal);
+
+  ps.add<bool>("reassignSeedCrysToClusterItSeeds", false);
```
Comment (Dr15Jones) [inline RecoEcal/EgammaClusterProducers/src/Multi5x5ClusterProducer.cc:106]: There should be a `iDesc.addDefault(ps)` here

**Comment by Dr15Jones** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by antoniovilela** [comment]
+1

**Comment by Dr15Jones** [comment]
@jfernan2 looking at the results of the PR test, it looks like this module is producing a non-negligible number of difference over the original code, e.g.

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-21-1100+2da7e6/65771/validateJR/1000.0_RunMinBias2011A/all_RECO_step2_inAOD/c_recoPreshowerClusters_multi5x5SuperClustersWithPreshower_preshowerXClusters_RECO_obj_eta.png

and

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-21-1100+2da7e6/65771/9.0_Higgs200ChargedTaus/EcalClusterV_EcalSuperClusters.html

My expectations were that the code changes should have given identical physics.

**Comment by makortel** [comment]
Should we revert the PR?

**Comment by Dr15Jones** [comment]
So I added a print statement into the module for both the original version and the new one. It appears the new one tends to have 1 more hit in clusters when things are different (below '<' denotes the new code and '>' the old code)

```diff
<  CaloCluster , algoID=4, CaloID: 2, E=28.5028, eta,phi=2.16838,0.610014, nhits=25
< ( 872440511, 1 ), ( 872440512, 1 ), ( 872440513, 1 ), ( 872440514, 1 ), ( 872440515, 1 ), ( 872440639, 1 ), ( 872440640, 1 ), ( 872440641, 1 ), ( 872440642, 1 ), ( 872440643, 1 ), ( 872440767, 1 ), ( 872440768, 1 ), ( 872440769, 1 ), ( 872440770, 1 ), ( 872440771, 1 ), ( 872440895, 1 ), ( 872440896, 1 ), ( 872440897, 1 ), ( 872440898, 1 ), ( 872440899, 1 ), ( 872441023, 1 ), ( 872441024, 1 ), ( 872441025, 1 ), ( 872441026, 1 ), ( 872441027, 1 ), 
<  CaloCluster , algoID=4, CaloID: 2, E=20.74, eta,phi=2.12595,0.736141, nhits=17
< ( 872440386, 1 ), ( 872440387, 1 ), ( 872440388, 1 ), ( 872440389, 1 ), ( 872440390, 1 ), ( 872440516, 1 ), ( 872440517, 1 ), ( 872440518, 1 ), ( 872440644, 1 ), ( 872440645, 1 ), ( 872440646, 1 ), ( 872440772, 1 ), ( 872440773, 1 ), ( 872440774, 1 ), ( 872440900, 1 ), ( 872440901, 1 ), ( 872440902, 1 ), 
---
>  CaloCluster , algoID=4, CaloID: 2, E=26.326, eta,phi=2.1625,0.614715, nhits=23
> ( 872440511, 1 ), ( 872440512, 1 ), ( 872440514, 1 ), ( 872440515, 1 ), ( 872440639, 1 ), ( 872440640, 1 ), ( 872440641, 1 ), ( 872440642, 1 ), ( 872440643, 1 ), ( 872440767, 1 ), ( 872440768, 1 ), ( 872440769, 1 ), ( 872440770, 1 ), ( 872440771, 1 ), ( 872440896, 1 ), ( 872440897, 1 ), ( 872440898, 1 ), ( 872440899, 1 ), ( 872441023, 1 ), ( 872441024, 1 ), ( 872441025, 1 ), ( 872441026, 1 ), ( 872441027, 1 ), 
>  CaloCluster , algoID=4, CaloID: 2, E=19.7252, eta,phi=2.1303,0.742665, nhits=16
> ( 872440386, 1 ), ( 872440387, 1 ), ( 872440388, 1 ), ( 872440389, 1 ), ( 872440390, 1 ), ( 872440516, 1 ), ( 872440517, 1 ), ( 872440518, 1 ), ( 872440644, 1 ), ( 872440645, 1 ), ( 872440646, 1 ), ( 872440772, 1 ), ( 872440773, 1 ), ( 872440774, 1 ), ( 872440901, 1 ), ( 872440902, 1 ), 
5576c5576
```
and 
```diff
<  CaloCluster , algoID=4, CaloID: 2, E=39.4991, eta,phi=2.76572,-1.41006, nhits=21
< ( 872438180, 1 ), ( 872438181, 1 ), ( 872438182, 1 ), ( 872438183, 1 ), ( 872438308, 1 ), ( 872438309, 1 ), ( 872438310, 1 ), ( 872438311, 1 ), ( 872438435, 1 ), ( 872438436, 1 ), ( 872438437, 1 ), ( 872438438, 1 ), ( 872438439, 1 ), ( 872438564, 1 ), ( 872438565, 1 ), ( 872438566, 1 ), ( 872438567, 1 ), ( 872438692, 1 ), ( 872438693, 1 ), ( 872438694, 1 ), ( 872438695, 1 ), 
---
>  CaloCluster , algoID=4, CaloID: 2, E=34.8612, eta,phi=2.74065,-1.40841, nhits=20
> ( 872438180, 1 ), ( 872438181, 1 ), ( 872438182, 1 ), ( 872438183, 1 ), ( 872438308, 1 ), ( 872438309, 1 ), ( 872438310, 1 ), ( 872438435, 1 ), ( 872438436, 1 ), ( 872438437, 1 ), ( 872438438, 1 ), ( 872438439, 1 ), ( 872438564, 1 ), ( 872438565, 1 ), ( 872438566, 1 ), ( 872438567, 1 ), ( 872438692, 1 ), ( 872438693, 1 ), ( 872438694, 1 ), ( 872438695, 1 ), 
5740c5740
```

and
```diff
5887,5888d5886
<  CaloCluster , algoID=4, CaloID: 2, E=7.5786, eta,phi=-2.53928,3.05121, nhits=6
< ( 872419380, 1 ), ( 872419507, 1 ), ( 872419508, 1 ), ( 872419637, 1 ), ( 872419765, 1 ), ( 872419891, 1 ), 
5891,5892d5888
<  CaloCluster , algoID=4, CaloID: 2, E=9.02588, eta,phi=-2.88095,-0.0599388, nhits=5
< ( 872423217, 1 ), ( 872423219, 1 ), ( 872423345, 1 ), ( 872423346, 1 ), ( 872423475, 1 ), 
5900a5897,5898
>  CaloCluster , algoID=4, CaloID: 2, E=5.76277, eta,phi=-2.56595,3.04665, nhits=5
> ( 872419507, 1 ), ( 872419508, 1 ), ( 872419637, 1 ), ( 872419765, 1 ), ( 872419891, 1 ), 
5902a5901,5902
>  CaloCluster , algoID=4, CaloID: 2, E=6.99991, eta,phi=-2.88841,-0.0345329, nhits=4
> ( 872423217, 1 ), ( 872423219, 1 ), ( 872423346, 1 ), ( 872423475, 1 ), 
5977c5977
```

**Pull request merged**